alarm/AlarmManager2.js

```diff
@@ -1602,4 +1459,167 @@ module.exports = class {
+    const e = this.createException(alarm, userFeedback);
+    if (!e) new Error("Unsupported Action!");
+    const related = alarms
+      .filter(relatedAlarm => e.match(relatedAlarm)).map(alarm => alarm.aid);
+    return related || []
+  }
+  createException(alarm, userFeedback, callback) {
+    let i_target = null;
+    let i_type = null;
+    //IGNORE
+    switch (alarm.type) {
+      case "ALARM_NEW_DEVICE":
+      case "ALARM_DEVICE_OFFLINE":
+      case "ALARM_DEVICE_BACK_ONLINE":
+        i_type = "mac"; // place holder, not going to be matched by any alarm/policy
+        i_target = alarm["p.device.mac"];
+        break;
+      case "ALARM_UPNP":
+        i_type = "devicePort";
+        if (userFeedback) {
+          switch (userFeedback.type) {
+            case "deviceAllPorts":
+              i_type = "deviceAllPorts";
+              break;
+            case "deviceAppPort":
+              i_type = "deviceAppPort";
+              break;
+            default:
+              // do nothing
+              break;
+          }
+        }
+        // policy should be created with mac
+        if (alarm["p.device.mac"]) {
+          i_target = util.format("%s:%s:%s",
+            alarm["p.device.mac"],
+            alarm["p.upnp.private.port"],
+            alarm["p.upnp.protocol"]
+          )
+        } else {
+          let targetIp = alarm["p.device.ip"];
+          dnsManager.resolveLocalHost(targetIp, (err, result) => {
+            if (err || result == null) {
+              log.error("Alarm doesn't have mac and unable to resolve ip:", targetIp, err);
+              callback && callback(new Error("Alarm doesn't have mac and unable to resolve ip:", targetIp));
+              return;
+            }
+            i_target = util.format("%s:%s:%s",
+              result.mac,
+              alarm["p.upnp.private.port"],
+              alarm["p.upnp.protocol"]
+            )
+          })
+        }
+        break;
+      default:
+        if (alarm["p.dest.name"] === alarm["p.dest.ip"]) {
+          i_type = "ip";
+          i_target = alarm["p.dest.ip"];
+        } else {
+          i_type = "dns";
+          i_target = alarm["p.dest.name"];
+        }
+        if (userFeedback) {
+          switch (userFeedback.type) {
+            case "domain":
+            case "dns":
+              i_type = "dns"
+              i_target = userFeedback.target
+              break
+            case "ip":
+              i_type = "ip"
+              i_target = userFeedback.target
+              break
+            case "category":
+              i_type = "category";
+              i_target = userFeedback.target;
+              break;
+            case "country":
+              i_type = "country";
+              i_target = userFeedback.target;
+              break;
+            default:
+              break
+          }
+        }
+        break;
+    }
+    if (userFeedback && userFeedback.archiveAlarmByType) {
```

Comment:
info.archiveAlarmByType & info.device are quite similar arguments but have different naming style. Make sure it's constant.